### ISSUE TITLE
Add option to request spot-instance with on-demand price

### ISSUE THREAD

**Issue by ilansh** [npalm/terraform-aws-gitlab-runner#222]
When leaving spot-price empty, the price defaults to on-demand price, no need to come up with arbitrary numbers.

https://aws.amazon.com/blogs/compute/new-amazon-ec2-spot-pricing/

**Comment by npalm** [comment]
@ilansh The module is only taking care of creating the infrastructure. Creating the ec2 instances is handled by the aws provider for docker machine. So as for I can see there is no option to support this feature in this module. It is the gitlab agent that request ec2 instances via docker machine. What is your view?

**Comment by ilansh** [comment]
Perhaps if docker-machine's `--amazonec2-spot-price` flag is left undefined.  
Not sure what's the default for it though.

**Comment by kitos9112** [comment]
This is an interesting topic, taking a look at the source code for the [AWS driver `docker-machine` ](https://github.com/docker/machine/blob/master/drivers/amazonec2/amazonec2.go), the default value for defaultSpotPrice is set to $0.5. This piece of code has not been recently updated though.

```
const (
	driverName                  = "amazonec2"
	ipRange                     = "0.0.0.0/0"
	machineSecurityGroupName    = "docker-machine"
	defaultAmiId                = "ami-c60b90d1"
	defaultRegion               = "us-east-1"
	defaultInstanceType         = "t2.micro"
	defaultDeviceName           = "/dev/sda1"
	defaultRootSize             = 16
	defaultVolumeType           = "gp2"
	defaultZone                 = "a"
	defaultSecurityGroup        = machineSecurityGroupName
	defaultSSHPort              = 22
	defaultSSHUser              = "ubuntu"
	defaultSpotPrice            = "0.50" <<-----------------------
	defaultBlockDurationMinutes = 0
)
```
I think it may be worth raising an issue in their side so we might have a developer taking an action and adopting the default sane AWS approach for matching the on-demand EC2 price instead of an arbitrary `0.5`

[AWS - API_RequestSpotInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RequestSpotInstances.html)

> SpotPrice - The maximum price per hour that you are willing to pay for a Spot Instance. The default is the On-Demand price.
> 
>     Type: String
> 
>     Required: No

**Comment by Fuuzetsu** [comment]
You can work around it by setting the price yourself, something like

```terraform
data "aws_pricing_product" "on_demand_job_instance" {
  provider     = aws.ap-south-1
  service_code = "AmazonEC2"

  filters {
    field = "location"
    value = data.aws_region.current.description
  }

  filters {
    field = "operatingSystem"
    value = "Linux"
  }

  filters {
    field = "instanceType"
    value = var.gitlab_task_instance_type
  }

  filters {
    field = "capacitystatus"
    value = "Used"
  }

  filters {
    field = "tenancy"
    value = "Shared"
  }

  filters {
    field = "preInstalledSw"
    value = "NA"
  }

  filters {
    field = "licenseModel"
    value = "No License required"
  }
}
```

then get the value with `tonumber(values(values(jsondecode(data.aws_pricing_product.on_demand_job_instance.result).terms.OnDemand)[0].priceDimensions)[0].pricePerUnit.USD)`; of course that's annoying and you'll have to be updating this if price changes but it's better than some default 0.50

**Comment by kayman-mk** [comment]
I think there is no need to do anything, right? You specify the maximum price you are willing to pay. If the current spot price is less, you get charged the lower price. If the current spot price is higher, you will not get a machine.

**Comment by Fuuzetsu** [comment]
>I think there is no need to do anything, right? You specify the maximum price you are willing to pay. If the current spot price is less, you get charged the lower price. If the current spot price is higher, you will not get a machine.

The runner forces you to provide a value otherwise it defaults to $0.50. This default isn't very useful, it's just a random number that doesn't reflect instance type(s) you've chosen: if you're using large instance type, you're never going to get it for $0.50.

So you're forced to pick a number at the moment and the most common scenario is to accept up to the on-demand price: as you say, the spot price is usually much less and the user ends up paying the lower price.

What's annoying to do is to look up this price for every instance type you need (something my workaround above does) and keep track of the price changing over time.

Basically, in order to accept any price up to on-demand which is what most people want, we have to specify no value and let AWS deal with it – and the runner doesn't let you do this, forcing you to pick a number up front and keep it updated. Most people don't have a hard-limit of say $2, they are willing to pay whatever it costs to get a machine but if they can pay less with spot, that's what they want.

**Comment by kayman-mk** [comment]
Ok, I tried this approach. Filtering the location with `aws_region.name` is not working for Europe as "EU Frankfurt" is expected and not "Europe (Frankfurt)". Any idea?
